### DISCUSSION TITLE
"mocking" tools/dependencies in `meson setup`?

### DISCUSSION THREAD

**Discussion by jjmaestro** [mesonbuild/meson#13822] [Q&A]
I'd like to gather information about a repo without having to fully build it.

I found the `meson-info/*.json` files (and `meson introspect`) quite useful, it's almost everything I need.

My problem is that the `meson-info` / introspection provides information about a configured build and I would like to have info about "all of the potential builds". That is, I'd like to be able to get the introspection info "forcing" certain build options and/or dependencies.

For example, say I have a main `meson.build` at the root of the project with:
```meson
(...)
ssl = dependency('openssl', required: false)
(...)
```
and a subdir with a `meson.build` that starts with
```meson
if not ssl.found()
  subdir_done()
endif
(...)
```

If I don't have `openssl` installed, then I won't be able to e.g. get any info for that subdir in the `intro-install_plan.json` or in `meson introspect`.

Thus, my question: 

**Is there a way to parse / evaluate the `meson.build` files directly? And/or run `meson setup` "mocking" some of the dependencies so that I could get all the info I need in `meson-info` / `meson introspect`?**

By mocking I mean making Meson assume a certain tool in a certain path exists so that `<DEP>.found()` will return `True`. Then, I'd have to build some script that turns whatever options / dependencies on-and-off.

I assume that this would work because `meson setup` won't be running anything / any tool except `meson` itself. Or am I wrong and `meson setup` can run whatever it wants?

Thanks!

**Comment by dcbaker** [comment]
At the very least Meson going to run any compilers you need and pkg-config.

The easiest way to for Meson to find dependencies is probably to replace pkg-config with something that will return found for the dependencies you want, and not found otherwise. Meson's interpreter works node by node, so Meson will never seen branches it doesn't go down at all, which makes what you want to do somewhat difficult to achieve, but probably with a some clever replacement of pkg-config and your compilers you could do it for most projects.

**Reply by jjmaestro** [reply]
Cool, thanks! I was checking [Dependency detection method](https://mesonbuild.com/Dependencies.html#dependency-detection-method) in the docs and there's more than one way although it definitely looks like `pkg-config` might cover most of the cases.

I'll see if I can come up with a replacement of sorts, thanks for the idea!
